### PR TITLE
Leave BUNDLE_GEMFILE unset in workers if it was unset in the master

### LINKED ISSUE #2133: BUNDLE_GEMFILE is incorrectly set to parent Gemfile in worker processes (when not setting BUNDLE_GEMFILE explicitly)
**Describe the bug**

Before #1893, if you used `prune_bundler` and started `bundle exec puma` without explicitly setting `BUNDLE_GEMFILE`, workers would not have a value set for the environment variable `BUNDLE_GEMFILE`. This is desirable when you want workers to use the Gemfile located in the application directory, especially if that directory is different from the directory that you started the puma master process in (such as with a phased restart).

After #1893, workers have `BUNDLE_GEMFILE` unexpectedly set to the path of the Gemfile used by the master process. I think this forces `bundler` in the workers to use that Gemfile instead of the one in the worker's working directory as it would if the environment variable was unset.

I think the intent of that PR was that if you start your server with `BUNDLE_GEMFILE=Gemfile.rails6 bundle exec puma`, that workers would have `BUNDLE_GEMFILE` set to the same value. But it ignores the fact that `Bundler` itself sets a value for `ENV['BUNDLE_GEMFILE']` even if it wasn't specified at startup time. Workers now end up inheriting this value, when they didn't before.

One consequence of this bug is that it is no longer possible to delete from your server the Gemfile that was used when you booted the puma master process. If you do, workers fail with the next phased restart. I imagine there are other bugs related to being able to change gems in the Gemfile for new versions of your application, but I haven't been able to reproduce those.

**Puma config:**

```
port 3000
workers 1
prune_bundler
directory '/usr/src/releases/current'

on_worker_boot do
  pp "BUNDLE_GEMFILE: #{ENV['BUNDLE_GEMFILE']}"
end
```

**To Reproduce**

I wrote a script that reproduces the error here: https://github.com/cjlarose/puma-phased-restart-could-not-find-gem-errors/tree/regression-1893 (I reused a repo from an unrelated issue)

Before #1893 , workers boot and print `BUNDLE_GEMFILE: `. After that PR, workers boot and print `BUNDLE_GEMFILE: /usr/src/releases/1582850222334837300/Gemfile`, referring to the path of the Gemfile that was used by the puma master process.

**Expected behavior**

If `BUNDLE_GEMFILE` was unspecified when you started the master process, it should remain unspecified in the worker processes.

**Desktop (please complete the following information):**
 - OS: Linux
 - Puma Version 8c9b3ebc7a86bdc5488704a0225b76409b9f555f

### PR DESCRIPTION
**Pull request by cjlarose**
### Description

Closes #2133

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: lib/puma/launcher.rb [hunk 1]
```diff
@@ -298,7 +298,8 @@ def prune_bundler
 
       log '* Pruning Bundler environment'
       home = ENV['GEM_HOME']
-      bundle_gemfile = ENV['BUNDLE_GEMFILE']
+      bundle_gemfile = ENV['BUNDLER_ORIG_BUNDLE_GEMFILE']
+      bundle_gemfile = nil if bundle_gemfile == 'BUNDLER_ENVIRONMENT_PRESERVER_INTENTIONALLY_NIL'
```
Comment (nateberkopec) [inline lib/puma/launcher.rb]: What's this about?

Comment (cjlarose) [reply lib/puma/launcher.rb]: Removed in 0b47735f78c370d103af5c1b8b581948f56cd688

**Comment by seven1m** [comment]
This is solid work. Love the test!

**Comment by nateberkopec** [comment]
I'm gonna play around with the test locally but the code itself LGTM

**Pull request merged**